### PR TITLE
Laboratory work #7, Sviridov Vyacheslav - 20FPL1

### PR DESCRIPTION
**Pull request by VSvir**

### CODE REVIEW

**Comment by SalnikovIgor** [comment]
Fix unittests

**Comment by SalnikovIgor** [comment]
Time to start the task at 6

**Comment by VSvir** [comment]
So, here's what I have for now. I don't understand where I should look for `max_context_length` property though and my PC feels bad while trying to get model summary on 64 batches as defined in tests files. Guess I'll discuss these issues with others tomorrow

**Comment by VSvir** [comment]
Oh, btw, should we check `output_shape` property by taking the size of the output layer?

File: lab_7_llm/settings.json [hunk 1]
```diff
@@ -1,8 +1,8 @@
...
   "parameters": {
-    "model": "",
-    "dataset": "",
-    "metrics": []
+    "model": "cointegrated/rubert-base-cased-nli-threeway",
+    "dataset": "xnli",
+    "metrics": ["accuracy"]
   },
-  "target_score": 0
+  "target_score": 4
```
Comment (SalnikovIgor) [inline lab_7_llm/settings.json]: I think it should change to 6?

**Review by SalnikovIgor — CHANGES_REQUESTED** [review]
Implement inference sample

**Comment by SalnikovIgor** [comment]
> So, here's what I have for now. I don't understand where I should look for `max_context_length` property though and my PC feels bad while trying to get model summary on 64 batches as defined in tests files. Guess I'll discuss these issues with others tomorrow

We take the `max_context_length` parameter from the model config. As for the inference with batch 64, I think if you have experiments with a smaller batch, then you should pass tests with a batch of 64

**Comment by SalnikovIgor** [comment]
> Oh, btw, should we check `output_shape` property by taking the size of the output layer?

I don't quite understand the question, we can discuss it at the seminar

File: lab_7_llm/settings.json (hunk 1 above)
Comment (VSvir) [reply lab_7_llm/settings.json]: Yes, I'll change it when I'll complete all tasks for this mark

**Comment by VSvir** [comment]
Thanks for replies, I've also asked my questions in chat, so I believe we'll figure everything out

File: lab_7_llm/start.py [hunk 2]
```diff
@@ -2,15 +2,43 @@
...
+    dataset = TaskDataset(preprocessor.data.head(100))
+
+    pipeline = LLMPipeline(configs['parameters']['model'], dataset, 120, 2, 'cpu')
+
+    model_analysis = pipeline.analyze_model()
+    print(model_analysis)
+
+    sample = dataset[randint(0, len(dataset))]
+    sample_inference = pipeline.infer_sample(sample)
+    print(f'SAMPLE: {sample}\nPREDICTION: {sample_inference}')
```
Comment (SalnikovIgor) [inline lab_7_llm/start.py]: This is information for debugging?

**Review by SalnikovIgor — CHANGES_REQUESTED** [review]
Delete the debug prints, wait for the references to be updated and fix the tests

File: lab_7_llm/start.py (hunk 2 above)
Comment (VSvir) [reply lab_7_llm/start.py]: Well, yes, but actually no
While I do use these prints for debugging, I was also planning to leave them here just to have a possibility to see the results of the previous steps which are required for lesser marks

Comment (VSvir) [reply lab_7_llm/start.py:38]: Actually, this is the purpose of all prints in `start.py`

**Comment by VSvir** [comment]
🎉

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+datasets==2.16.1
+evaluate==0.4.1
+numpy==1.26.3
+pandas==2.1.4
+pydantic==2.5.3
+pytest==7.4.4
+scikit-learn==1.4.0
+torch==2.1.2
+torchinfo==1.8.0
+tqdm==4.66.1
```
Comment (SalnikovIgor) [inline requirements.txt]: Also, why do we need `tqdm`?

File: requirements.txt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+--extra-index-url https://download.pytorch.org/whl/cpu
+
+datasets==2.16.1
+evaluate==0.4.1
+numpy==1.26.3
+pandas==2.1.4
+pydantic==2.5.3
```
Comment (SalnikovIgor) [inline requirements.txt:8]: Why do we need `pydantic`?

**Review by SalnikovIgor — CHANGES_REQUESTED** [review]
Fix minor comments and when do you plan to implement a service for a score of 10?

**Comment by SalnikovIgor** [comment]
Also don't forget to remove debbug prints

**Comment by VSvir** [comment]
Oh, looks like we've broken the website

**Comment by VSvir** [comment]
> when do you plan to implement a service for a score of 10?

image

**Comment by VSvir** [comment]
Bruh 🗿

**Comment by VSvir** [comment]
🎉🎉🎉

**Comment by VSvir** [comment]
That's it 👍🏻

**Comment by VSvir** [comment]
Now that's actually it

File: lab_7_llm/assets/main.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    await fetch('/infer', {
+        method: 'POST',
+        headers: {
+            'Content-type': 'application/json'
+        },
+        body: JSON.stringify({
+            question: premise.value,
+            hypothesis: hypothesis.value
+        })
+    }).then(res => {return res.json()})
```
Comment (SalnikovIgor) [inline lab_7_llm/assets/main.js]: I think this is the third time I've seen this code.
`async \ await` function was invented as the development of promises and getting rid of `then` and promise hell

File: lab_7_llm/main.py [hunk 6]
```diff
@@ -32,6 +27,7 @@ class RawDataImporter(AbstractRawDataImporter):
     """
     A class that imports the HuggingFace dataset.
     """
+    _raw_data: DataFrame
```
Comment (SalnikovIgor) [inline lab_7_llm/main.py:30]: This parameter is already created in the class `AbstractRawDataImporter`

File: lab_7_llm/start.py [hunk 7]
```diff
@@ -2,15 +2,49 @@
...
+    sample = dataset[randint(0, len(dataset) - 1)]
+    sample_inference = pipeline.infer_sample(sample)
+
+    predictions = pipeline.infer_dataset()
+    data_path = PROJECT_ROOT / 'lab_7_llm' / 'dist' / 'predictions.csv'
+    pipeline.save_results(predictions, data_path)
+
+    evaluator = TaskEvaluator(
+        data_path,
+        [Metrics[metric.upper()] for metric in configs['parameters']['metrics']]
```
Comment (SalnikovIgor) [inline lab_7_llm/start.py:44]: Do you think can we  pass this parameter differently?

**Review by SalnikovIgor — CHANGES_REQUESTED** [review]

File: lab_7_llm/main.py (hunk 6 above)
Comment (VSvir) [reply lab_7_llm/main.py:30]: Yes, but without overriding a type of this parameter Mypy checks were failing

File: lab_7_llm/start.py (hunk 7 above)
Comment (VSvir) [reply lab_7_llm/start.py:44]: I guess so, probably we could pass a whole Enum here and then look for a needed metric in `TaskEvaluator`, but this approach needs us to open a settings file in `main.py` module which I would prefer to refrain from

**Pull request closed without merging**